File: mocida/src/uikit/container.c
#include <uikit/container.h>
#include <uikit/rect.h>
#include <stdlib.h>
#include <string.h>

// =====================================================================
// UIGrid
// =====================================================================

UIGrid* UIGrid_Create(int columns) {
    if (columns < 1) columns = 1;
    UIGrid* g = (UIGrid*)calloc(1, sizeof(UIGrid));
    if (!g) return NULL;
    g->__widget_type = UI_WIDGET_GRID;
    g->columns = columns;
    g->gapX = g->gapY = 8.0f;
    g->cellW = g->cellH = 100.0f;
    g->items = UIChildren_Create(16);
    if (!g->items) { free(g); return NULL; }
    return g;
}

UIGrid* UIGrid_SetColumns(UIGrid* g, int columns) {
    if (!g) return g;
    if (columns < 1) columns = 1;
    g->columns = columns;
    return g;
}

UIGrid* UIGrid_SetGap(UIGrid* g, float gapX, float gapY) {
    if (!g) return g;
    g->gapX = gapX;
    g->gapY = gapY;
    return g;
}

UIGrid* UIGrid_SetCellSize(UIGrid* g, float w, float h) {
    if (!g) return g;
    g->cellW = w;
    g->cellH = h;
    return g;
}

UIGrid* UIGrid_SetPadding(UIGrid* g, float l, float t, float r, float b) {
    if (!g) return g;
    g->paddingLeft = l;
    g->paddingTop = t;
    g->paddingRight = r;
    g->paddingBottom = b;
    return g;
}

int UIGrid_AddItem(UIGrid* g, UIWidget* item) {
    if (!g || !item) return 0;
    return UIChildren_Add(g->items, item);
}

void UIGrid_GetContentSize(UIGrid* g, float* outW, float* outH) {
    if (!g) { if (outW) *outW = 0; if (outH) *outH = 0; return; }
    const int count = g->items ? g->items->count : 0;
    if (count == 0) {
        if (outW) *outW = g->paddingLeft + g->paddingRight;
        if (outH) *outH = g->paddingTop + g->paddingBottom;
        return;
    }
    const int cols = g->columns;
    const int rows = (count + cols - 1) / cols;
    const float w = g->paddingLeft + g->paddingRight
                  + cols * g->cellW + (cols - 1) * g->gapX;
    const float h = g->paddingTop + g->paddingBottom
                  + rows * g->cellH + (rows - 1) * g->gapY;
    if (outW) *outW = w;
    if (outH) *outH = h;
}

void UIGrid_Destroy(UIGrid* g) {
    if (!g) return;
    if (g->items) UIChildren_Destroy(g->items);
    free(g);
}

// =====================================================================
// UIScroll
// =====================================================================

UIScroll* UIScroll_Create(void) {
    UIScroll* s = (UIScroll*)calloc(1, sizeof(UIScroll));
    if (!s) return NULL;
    s->__widget_type = UI_WIDGET_SCROLL;
    s->allowVertical = 1;
    s->allowHorizontal = 1;
    s->allowDragScroll = 0;
    s->wheelSpeed = 60.0f;
    return s;
}

UIScroll* UIScroll_SetContent(UIScroll* s, UIWidget* content) {
    if (!s) return s;
    if (s->content) {
        UIWidget_Destroy(s->content);
    }
    s->content = content;
    // Invalidate cached size.
    s->contentW = 0.0f;
    s->contentH = 0.0f;
    return s;
}

UIScroll* UIScroll_SetScroll(UIScroll* s, float x, float y) {
    if (!s) return s;
    s->scrollX = x;
    s->scrollY = y;
    return s;
}

UIScroll* UIScroll_SetAxes(UIScroll* s, int allowVertical, int allowHorizontal) {
    if (!s) return s;
    s->allowVertical = allowVertical ? 1 : 0;
    s->allowHorizontal = allowHorizontal ? 1 : 0;
    return s;
}

UIScroll* UIScroll_SetDragScroll(UIScroll* s, int enabled) {
    if (!s) return s;
    s->allowDragScroll = enabled ? 1 : 0;
    return s;
}

UIScroll* UIScroll_SetWheelSpeed(UIScroll* s, float pxPerNotch) {
    if (!s) return s;
    s->wheelSpeed = pxPerNotch;
    return s;
}

void UIScroll_InvalidateContentSize(UIScroll* s) {
    if (!s) return;
    s->contentW = 0.0f;
    s->contentH = 0.0f;
}

void UIScroll_Destroy(UIScroll* s) {
    if (!s) return;
    if (s->content) UIWidget_Destroy(s->content);
    if (s->background) UIRectangle_Destroy((UIRectangle*)s->background);
    free(s);
}

// =====================================================================
// UIListView (1-column grid wrapped in a scroll)
// =====================================================================

UIScroll* UIListView_Create(float itemHeight) {
    UIScroll* s = UIScroll_Create();
    if (!s) return NULL;
    UIScroll_SetAxes(s, /*vert=*/1, /*horiz=*/0);

    UIGrid* g = UIGrid_Create(1);
    UIGrid_SetCellSize(g, 0.0f, itemHeight > 0.0f ? itemHeight : 60.0f);
    UIGrid_SetGap(g, 0.0f, 6.0f);
    UIGrid_SetPadding(g, 8.0f, 8.0f, 8.0f, 8.0f);

    // The cellW will be replaced with the scroll's bounds at first
    // render (see RenderWidget in window.c). Stored 0 means "stretch
    // to viewport width".
    UIWidget* gridW = UIWidget_Create(g);
    UIScroll_SetContent(s, gridW);
    return s;
}

UIGrid* UIListView_GetGrid(UIScroll* listView) {
    if (!listView || !listView->content) return NULL;
    UIWidgetBase* base = (UIWidgetBase*)listView->content->data;
    if (!base || strcmp(base->__widget_type, UI_WIDGET_GRID) != 0) return NULL;
    return (UIGrid*)base;
}

int UIListView_AddItem(UIScroll* listView, UIWidget* item) {
    UIGrid* g = UIListView_GetGrid(listView);
    if (!g) return 0;
    return UIGrid_AddItem(g, item);
}

// =====================================================================
// UIGridView (N-column grid wrapped in a scroll)
// =====================================================================

UIScroll* UIGridView_Create(int columns, float cellW, float cellH) {
    UIScroll* s = UIScroll_Create();
    if (!s) return NULL;
    UIScroll_SetAxes(s, /*vert=*/1, /*horiz=*/0);

    UIGrid* g = UIGrid_Create(columns);
    UIGrid_SetCellSize(g, cellW, cellH);
    UIGrid_SetGap(g, 10.0f, 10.0f);
    UIGrid_SetPadding(g, 12.0f, 12.0f, 12.0f, 12.0f);

    UIWidget* gridW = UIWidget_Create(g);
    UIScroll_SetContent(s, gridW);
    return s;
}

UIGrid* UIGridView_GetGrid(UIScroll* gridView) {
    return UIListView_GetGrid(gridView); // same shape
}

int UIGridView_AddItem(UIScroll* gridView, UIWidget* item) {
    return UIListView_AddItem(gridView, item);
}

File: mocida-rs/mocida/src/container.rs
//! `UIGrid` and `UIScroll` containers (plus `ListView` / `GridView` sugar).

use mocida_sys as sys;

use crate::error::{Error, Result};
use crate::widget::Widget;

/// Fixed-cell grid container. Lays children row-major from the
/// top-left into `columns` columns of `cell_w x cell_h` cells.
pub struct Grid {
    ptr: *mut sys::UIGrid,
    moved: bool,
}

impl Grid {
    /// Creates a grid with the given column count.
    pub fn new(columns: i32) -> Result<Self> {
        let ptr = unsafe { sys::UIGrid_Create(columns) };
        if ptr.is_null() {
            return Err(Error::Null("UIGrid_Create"));
        }
        Ok(Self { ptr, moved: false })
    }

    /// Updates the column count.
    pub fn columns(self, columns: i32) -> Self {
        unsafe { sys::UIGrid_SetColumns(self.ptr, columns) };
        self
    }

    /// Sets the horizontal and vertical gap between cells.
    pub fn gap(self, gap_x: f32, gap_y: f32) -> Self {
        unsafe { sys::UIGrid_SetGap(self.ptr, gap_x, gap_y) };
        self
    }

    /// Sets the cell dimensions.
    pub fn cell_size(self, width: f32, height: f32) -> Self {
        unsafe { sys::UIGrid_SetCellSize(self.ptr, width, height) };
        self
    }

    /// Sets inner padding (left, top, right, bottom).
    pub fn padding(self, left: f32, top: f32, right: f32, bottom: f32) -> Self {
        unsafe { sys::UIGrid_SetPadding(self.ptr, left, top, right, bottom) };
        self
    }

    /// Appends an item. The grid takes ownership of the widget.
    pub fn add(&mut self, item: Widget) -> Result<()> {
        let raw = item.into_raw();
        let rc = unsafe { sys::UIGrid_AddItem(self.ptr, raw) };
        if rc != 1 {
            return Err(Error::Null("UIGrid_AddItem"));
        }
        Ok(())
    }

    /// Returns the total content `(width, height)` reported by mocida.
    pub fn content_size(&self) -> (f32, f32) {
        let mut w = 0.0f32;
        let mut h = 0.0f32;
        unsafe { sys::UIGrid_GetContentSize(self.ptr, &mut w, &mut h) };
        (w, h)
    }

    /// Borrow the raw `UIGrid*`.
    #[inline]
    pub fn as_ptr(&self) -> *mut sys::UIGrid {
        self.ptr
    }

    /// Lift into a [`Widget`] (auto-sized).
    pub fn into_widget(mut self) -> Result<Widget> {
        self.moved = true;
        unsafe { Widget::new(self.ptr as *mut _) }
    }

    /// Lift into a [`Widget`] with an explicit size.
    pub fn into_widget_sized(mut self, width: f32, height: f32) -> Result<Widget> {
        self.moved = true;
        unsafe { Widget::with_size(self.ptr as *mut _, width, height) }
    }
}

impl Drop for Grid {
    fn drop(&mut self) {
        if !self.moved && !self.ptr.is_null() {
            unsafe { sys::UIGrid_Destroy(self.ptr) };
        }
    }
}

/// Scrolling viewport widget with a single owned content child.
pub struct Scroll {
    ptr: *mut sys::UIScroll,
    moved: bool,
}

impl Scroll {
    /// Allocates a new scroll viewport.
    pub fn new() -> Result<Self> {
        let ptr = unsafe { sys::UIScroll_Create() };
        if ptr.is_null() {
            return Err(Error::Null("UIScroll_Create"));
        }
        Ok(Self { ptr, moved: false })
    }

    /// Sets the inner content. The scroll takes ownership of the
    /// widget.
    pub fn content(self, widget: Widget) -> Self {
        let raw = widget.into_raw();
        unsafe { sys::UIScroll_SetContent(self.ptr, raw) };
        self
    }

    /// Sets the scroll offsets.
    pub fn scroll(self, x: f32, y: f32) -> Self {
        unsafe { sys::UIScroll_SetScroll(self.ptr, x, y) };
        self
    }

    /// Enables / disables vertical and horizontal scrolling.
    pub fn axes(self, vertical: bool, horizontal: bool) -> Self {
        unsafe { sys::UIScroll_SetAxes(self.ptr, vertical as i32, horizontal as i32) };
        self
    }

    /// Enables drag-to-pan with the left mouse button.
    pub fn drag_scroll(self, enabled: bool) -> Self {
        unsafe { sys::UIScroll_SetDragScroll(self.ptr, enabled as i32) };
        self
    }

    /// Sets the wheel scroll speed in pixels per notch.
    pub fn wheel_speed(self, px_per_notch: f32) -> Self {
        unsafe { sys::UIScroll_SetWheelSpeed(self.ptr, px_per_notch) };
        self
    }

    /// Invalidates the cached content size.
    pub fn invalidate_content_size(&mut self) {
        unsafe { sys::UIScroll_InvalidateContentSize(self.ptr) };
    }

    /// Borrow the raw `UIScroll*`.
    #[inline]
    pub fn as_ptr(&self) -> *mut sys::UIScroll {
        self.ptr
    }

    /// Lift into a [`Widget`] (auto-sized).
    pub fn into_widget(mut self) -> Result<Widget> {
        self.moved = true;
        unsafe { Widget::new(self.ptr as *mut _) }
    }

    /// Lift into a [`Widget`] with an explicit size.
    pub fn into_widget_sized(mut self, width: f32, height: f32) -> Result<Widget> {
        self.moved = true;
        unsafe { Widget::with_size(self.ptr as *mut _, width, height) }
    }
}

impl Drop for Scroll {
    fn drop(&mut self) {
        if !self.moved && !self.ptr.is_null() {
            unsafe { sys::UIScroll_Destroy(self.ptr) };
        }
    }
}

/// Vertical list view (scroll + single-column grid). Mirrors
/// `UIListView_Create`.
pub struct ListView {
    inner: Scroll,
}

impl ListView {
    /// Creates a vertical list with rows of `item_height`.
    pub fn new(item_height: f32) -> Result<Self> {
        let ptr = unsafe { sys::UIListView_Create(item_height) };
        if ptr.is_null() {
            return Err(Error::Null("UIListView_Create"));
        }
        Ok(Self {
            inner: Scroll {
                ptr,
                moved: false,
            },
        })
    }

    /// Appends an item to the list. The list owns the widget.
    pub fn add(&mut self, item: Widget) -> Result<()> {
        let raw = item.into_raw();
        let rc = unsafe { sys::UIListView_AddItem(self.inner.ptr, raw) };
        if rc != 1 {
            return Err(Error::Null("UIListView_AddItem"));
        }
        Ok(())
    }

    /// Borrow the underlying [`Scroll`].
    pub fn as_scroll(&self) -> &Scroll {
        &self.inner
    }

    /// Consume into the underlying [`Scroll`].
    pub fn into_scroll(self) -> Scroll {
        self.inner
    }

    /// Convenience: lift the list view directly into a [`Widget`].
    pub fn into_widget_sized(self, width: f32, height: f32) -> Result<Widget> {
        self.inner.into_widget_sized(width, height)
    }
}

/// Two-dimensional grid view (scroll + grid).
pub struct GridView {
    inner: Scroll,
}

impl GridView {
    /// Creates a grid view with the given column count and cell size.
    pub fn new(columns: i32, cell_w: f32, cell_h: f32) -> Result<Self> {
        let ptr = unsafe { sys::UIGridView_Create(columns, cell_w, cell_h) };
        if ptr.is_null() {
            return Err(Error::Null("UIGridView_Create"));
        }
        Ok(Self {
            inner: Scroll {
                ptr,
                moved: false,
            },
        })
    }

    /// Appends an item.
    pub fn add(&mut self, item: Widget) -> Result<()> {
        let raw = item.into_raw();
        let rc = unsafe { sys::UIGridView_AddItem(self.inner.ptr, raw) };
        if rc != 1 {
            return Err(Error::Null("UIGridView_AddItem"));
        }
        Ok(())
    }

    /// Convenience: lift directly into a [`Widget`].
    pub fn into_widget_sized(self, width: f32, height: f32) -> Result<Widget> {
        self.inner.into_widget_sized(width, height)
    }
}
